pkg/datastore/headerscontroller.go

```diff
@@ -0,0 +1,282 @@
+	defer hc.cl.RUnlock()
+	if len(hc.ordered) == 0 {
+		return nil, false
+	}
+
+	l := hc.ordered[len(hc.ordered)-1]
+	for _, v := range hc.ordered {
+		if !(l.Slot-v.Slot > InMemorySlotLag && time.Since(v.Added) > InMemorySlotTimeLag) {
+			return toBeRemoved, ok
+		}
+		toBeRemoved = append(toBeRemoved, v.Slot)
+		ok = true
+	}
+
+	return toBeRemoved, ok
+
+}
+
+func (hc *HeaderController) GetOrderedDesc() (info []uint64) {
+	hc.cl.RLock()
+	defer hc.cl.RUnlock()
+	info = make([]uint64, len(hc.ordered))
+	for i := len(hc.ordered) - 1; i >= 0; i-- {
+		info = append(info, hc.ordered[i].Slot)
+	}
+
+	return info
+}
+
+// addToOrdered to be run in lock
+func (hc *HeaderController) addToOrdered(i Info) {
+	hc.ordered = append(hc.ordered, i)
+	l := len(hc.ordered)
+	if l > 1 {
+		o := hc.ordered[l-1]
+		if o.Slot >= i.Slot {
+			sort.Slice(hc.ordered, func(i, j int) bool {
+				return hc.ordered[i].Slot < hc.ordered[j].Slot
+			})
+		}
+	}
+}
+
+func (hc *HeaderController) GetLatestSlot() (slot uint64) {
+	return atomic.LoadUint64(hc.latestSlot)
+}
+
+func (hc *HeaderController) AddMultiple(slot uint64, hnt []structs.HeaderAndTrace) (err error) {
+	hc.cl.Lock()
+	defer hc.cl.Unlock()
+
+	h, ok := hc.content[slot]
+	if !ok {
+		h = NewHNTs()
+		hc.content[slot] = h
+		hc.addToOrdered(Info{
+			Slot:  slot,
+			Added: time.Now(),
+		})
+
+		ls := atomic.LoadUint64(hc.latestSlot)
+		if ls == 0 || slot > ls {
+			atomic.StoreUint64(hc.latestSlot, slot)
+		}
+	}
+
+	for _, s := range hnt {
+		if err := h.AddContent(s); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (hc *HeaderController) Add(slot uint64, hnt structs.HeaderAndTrace) (newCreated bool, err error) {
+	hc.cl.Lock()
+	defer hc.cl.Unlock()
+
+	h, ok := hc.content[slot]
+	if !ok {
+		h = NewHNTs()
+		hc.content[slot] = h
+		hc.addToOrdered(Info{
+			Slot:  slot,
+			Added: time.Now(),
+		})
+		newCreated = true
+
+		ls := atomic.LoadUint64(hc.latestSlot)
+		if ls == 0 || slot > ls {
+			atomic.StoreUint64(hc.latestSlot, slot)
+		}
+	}
+	return newCreated, h.AddContent(hnt)
+}
+
+func (hc *HeaderController) GetContent(startingSlot, stopSlot uint64, limit int) (elements []structs.HeaderAndTrace, lastSlot uint64) {
+	for _, o := range hc.GetOrderedDesc() {
+		if o > startingSlot || o < stopSlot {
+			continue
+		}
+		hc.cl.RLock()
+		h, ok := hc.content[o]
+		hc.cl.RUnlock()
+		if !ok || h == nil {
+			continue
+		}
+		lastSlot = o
+		c, _ := h.GetContent()
+		elements = append(elements, c...)
+		if len(elements) >= limit {
+			return elements, lastSlot
+		}
+	}
+
+	return elements, lastSlot
+}
+
+func (hc *HeaderController) GetSingleSlot(slot uint64) (elements []structs.HeaderAndTrace, revision uint64, err error) {
+	hc.cl.RLock()
+	h, ok := hc.content[slot]
+	hc.cl.RUnlock()
+
+	if !ok {
+		return nil, 0, nil
+	}
+	elements, revision = h.GetContent()
+	return elements, revision, err
+}
+
+func (hc *HeaderController) UnlinkElement(slot, expectedRevision uint64) (success bool) {
```

Comment:
`UnlinkElement` is accurate, but for giving a more descriptive name in this context, I would call it something similar to `deleteFromMemory`

Reply:
Delete from memory happens later by GC. This is just unlinking this from the map. Maybe we can find some better name...

Reply:
Yes, yes, delete is not accurate enough. What about something like `UnmapHeader`, or something on that direction?

Reply:
`RemoveSlot` ?

Reply:
Good for me

Reply:
done